### DISCUSSION TITLE
Automatically remove unused imports

### DISCUSSION THREAD

**Discussion by 11Firefox11** [biomejs/biome#305] [Ideas]
Have an option in [`organizeImports`](https://biomejs.dev/reference/configuration/#organizeimports) to remove imports that aren't used. 

From:
```ts
import { AfterViewInit, Component, ElementRef, Input, QueryList, ViewChildren } from "@angular/core";

@Component({
  selector: "app-my",
  templateUrl: "./my.component.html",
})
export class MyComponent {
  @Input() title: string;
}
```

To:
```ts
import { Component, Input } from "@angular/core";

@Component({
  selector: "app-my",
  templateUrl: "./my.component.html",
})
export class MyComponent {
  @Input() title: string;
}
```

An idea discussion based on question #304 .

**Comment by Conaclos** [comment]
Could you add more details? Which functionality automatically removes unused imports, ...

**Reply by 11Firefox11** [reply]
Added where it should be and a basic example.

**Reply by Conaclos** [reply]
Thanks! Let's take some time to allow peoples to share their opinion. By the way, I think we can remove the option. This should be the default behavior of the command.

**Reply by wcastand** [reply]
i wouldn't be against it but not as the default, when i code/debug, i have often unused variable / imports hanging around because i delete/add code to debug. i had the experience of my linter removing those every time each save and it was a pain in the ass, waste of time. So i don't think we should have this as the default behaviour. 

But something that can be enable as cli option (for example to be able to make it part of the husky or ci format would be nice)

**Reply by Conaclos** [reply]
I think it could be worth to split `noUnusedVariables` into two rules: `noUnusedImports` and `noUnusedVariables`. We should also support different workflows:

- removing unused imports on save
- removing unused imports later
- just warn about unused imports

To support these different workflows, I opened a broader discussion #368 because this could benefit to other rules.

**Comment by ematipico** [comment]
I believe import sorting should stick to that: moving the code around, but not eliminating it.

Removing code is more for a linter, where the user **chooses** to do it.

**Reply by Conaclos** [reply]
Could this also be a LSP action `Remove unused imports`?

We could have both: a linter rule and a LSP action?

**Reply by ematipico** [reply]
Not sure it makes sense to have a lint rule and an action rule... it should be one or the other

**Reply by Conaclos** [reply]
`Organize imports` is both usable on the LSP and on the CLI (via the config). Do you think it makes sense to do the same for `Remove unused imports`? If no, then I think we should stick with the linter rule in order to allow its use on CLI.

**Reply by ematipico** [reply]
Organize import **is not** a lint rule, but it's an **action** rule 😅

It's threaded differently. And that's why import sorting isn't part of the lint phase. It has it's own phase.

People come from eslint, and they are taught that unused imports is a lint rule. So it's about expectations.

If we want this rule to behave like import sorting, then it will be an action rule and you won't get the red marker in the IDE.

**Reply by wcastand** [reply]
i think i would like it to be an action where it doesn't throw in the editor so i can hack and debug freely but have it run on commit and ci to clean it up once i'm finished.

**Comment by 11Firefox11** [comment]
Is there any update on this?

**Reply by ematipico** [reply]
https://biomejs.dev/linter/rules/no-unused-imports/#_top

**Reply by 11Firefox11** [reply]
Thank you so much guys!

**Reply by wcastand** [reply]
altho the `--apply` doesn't delete the line.

so i end up with 

```
import ...
import ...

import ...
import ...
```

where i'd like 
```
import ...
import ...
import ...
import ...
```

**Reply by Conaclos** [reply]
@wcastand Could you open an issue (bug issue) with a playground link to reproduce the problem?

**Reply by wcastand** [reply]
@Conaclos can't manage to make the import rule fix to be ran. where do i enable the fix to be ran?
i get the warning by having all rules.

[playground](https://biomejs.dev/playground/?lintRules=all&code=aQBtAHAAbwByAHQAIAB7AGEAfQAgAGYAcgBvAG0AIAAnAGEAJwAKAGkAbQBwAG8AcgB0ACAAewBkAH0AIABmAHIAbwBtACAAJwBkACcACgBpAG0AcABvAHIAdAAgAHsAYgB9ACAAZgByAG8AbQAgACcAYgAnAAoAaQBtAHAAbwByAHQAIAB7AGMAfQAgAGYAcgBvAG0AIAAnAGMAJwAKAGkAbQBwAG8AcgB0ACAAewBlAH0AIABmAHIAbwBtACAAJwBlACcACgAKAGUAeABwAG8AcgB0ACAAYwBvAG4AcwB0ACAAYgBiACAAPQAgAGEAIAArACAAYgAgACsAIABjACAAKwAgAGUACgAKAA%3D%3D)

**Reply by Conaclos** [reply]
@wcastand It is not possble. You just get the fix suggestion in console. This is enough :)

**Reply by wcastand** [reply]
https://github.com/biomejs/biome/issues/653

**Reply by unional** [reply]
Want to share that if the code contains those empty lines to begin with, e.g.:

```ts
import ...
import ...

import ...
```

Those empty lines should not be deleted.
I have seen legit use cases of these empty lines in various projects.
e.g.

```ts
import { code } from 'some-code'

import json from './some.json'
```

```ts
import { code } from 'some-code'

impot { type Type } from './some-code.ts'
```

```ts
import { code } from 'some-code'

import { local } from './local.js'
```

**Comment by Conaclos** [comment]
We now provide the `noUnusedImports` rule.

**Comment by vimutti77** [comment]
My use case are
- Format but don't remove anything on save
- Format and remove unimports  and unused variables when I press the format button (Shift+Alt+F)

How can I achieve this?

**Comment by kravetsone** [comment]
I guess this should be implemented. But possible by flag behaviour 

I am big fan of fix-on-save but if it will erase imports which i type on this moment it would be evil

So something like

`bunx @biomejs/biome lint --write --remove-unused-imports` would be fine

**Reply by Conaclos** [reply]
You can already do something like that [using `--only=rule`](https://biomejs.dev/linter/#run-a-rule-or-a-group)

`bunx @biomejs/biome lint --write --only=correctness/noUnusedImports`

**Reply by kravetsone** [reply]
> You can already do something like that [using `--only=rule`](https://biomejs.dev/linter/#run-a-rule-or-a-group)
> 
> `bunx @biomejs/biome lint --write --only=correctness/noUnusedImports`

Oh sorry 
That's great!

**Comment by AlfianChabib** [comment]
<img width="344" height="119" alt="image" src="https://github.com/user-attachments/assets/eb9996cd-ba2c-4fdf-ad8b-6c6eacef0332" />
